### ISSUE TITLE
bug: capturing mouse

### ISSUE THREAD

**Issue by drcmda** [omgovich/react-colorful#97]
it looks like react-colorful does not capture, that easily breaks controls which depend on drag, simply by moving the mouse out of its region or the window. i've tried to use this lib (for instance in the use-tweaks library) but it constantly breaks and feels unstable bc of this.

i have a small repro here: https://codesandbox.io/s/floating-shoe-forked-qxjoj

if you click a segment of the shoe it shows the color picker, drag a color outside the window and let go (mouse-up), if you come in again the picker is still in drag mode. it also happens sometimes if you let go at the edge of the picker.

capturing will solve all these issues, would you add it?

**Comment by drcmda** [comment]
i believe this should do it...

```jsx
  const handleMoveStart = useCallback(
    ({ nativeEvent: event }: React.MouseEvent | React.TouchEvent) => {
      event.preventDefault();
      event.target.setPointerCapture(event.pointerId);
```

**Comment by omgovich** [comment]
Hi @drcmda!
Thanks, I'll release a patch soon!

**Comment by omgovich** [comment]
**react-colorful** aims to support legacy browsers like IE11, so I'll have to spend some time double-checking that using PE won't break anything. But I like the idea. Hope PE works okay everywhere.

**Comment by omgovich** [comment]
`PointerEvent` support was added to Safari since v13 (mid 2019) so It seems to me that I have to add a fallback for `mouse*` (desktop Safari) and `touch*` (iOS Safari) events anyway.

**Comment by drcmda** [comment]
i think the main issue is that it's not working right now. it hangs up pretty much every time i drag the slider and then goes into undefined mode where it's hanging on to an old state. imo if some corners have to be cut i'd rather throw some old browsers under the bus instead of coping with somewhat broken behaviour. i see more and more libs going PE, should be alright. or is there another way to fix it? this is something that react-color for instance does well - and im not sure if they really use PE so perhaps they have found another solution?

**Comment by dbismut** [comment]
@drcmda it works fine on my laptop (MacBook Pro / Chrome).

**Comment by omgovich** [comment]
Checked some libs. Examples:
**react-color** uses mouse event only (no PE) — it doesn't work well on mobile.
 **react-beautiful-dnd** works with mouse and touch events. No PE.

All these libs (as **react-colorful** do) work well on desktop without PE until the component is **not** placed inside the iframe (the way the codepen preview works). If you move the cursor outside of the regular window — everything is okay. If component is inside the iframe — regular mouse- and touch-events don't work as expected.

I already made the PR replacing mouse and touch events  with the pointer ones (#98), but I still looking a good solution to keep Safari <13 support (hope it's exist)

**Comment by omgovich** [comment]
Hi @dbismut! As far as I see react-use-gesture works well inside of codesandbox's `iframe`. How do you achieve that? PointrEvents or something else?

**Comment by drcmda** [comment]
@dbismut sure about that? i use a mbp/chrome and the behaviour is expected. without capture the mouse event must break if i go outside the window and then lift fingers (mouse-up). if i go back into the window the control is now broken and thinks it's still being dragged.

**Comment by dbismut** [comment]
I wasn't testing in an iframe so maybe I'm off topic, but yes it works fine in normal situation,I tried dragging off the window bounds and the drag persisted.

**Comment by drcmda** [comment]
no, that's not what i mean, it persists, but lift your fingers off - then go back into the window.

**Comment by dbismut** [comment]
Oooh right. That will teach me not to read issues with enough focus.

So in react use gesture we now use Pointer Events, but in earlier versions I tackled the problem by checking if event.buttons was not 0. If it was then I fired onMouseUp.

**Comment by dbismut** [comment]
https://github.com/pmndrs/react-use-gesture/blob/v7/src/recognizers/DragRecognizer.ts#L117-120

**Comment by omgovich** [comment]
@dbismut What about Safari 12 support? PonterEvent is not supported there. have you received complaints from developers about missing Safari 12 support? I'm afraid that Safari 13 was released in May 2019 and there are users (4% of the market actually) that use Safari <13.

**Comment by dbismut** [comment]
@omgovich

- In React Use Gesture (<8) we had an option to use PointerEvent.
- But by default we used MouseEvent and TouchEvent.
- In v8, we ditched MouseEvent and TouchEvent from the drag gesture and switched to PointerEvent. These were still attached to the window.
- In v9 (which should be released by end of the year), we're going to use `setPointerCapture` and remove the need to attach listeners to the window.

At this point I didn't received any complaints, but in any case people could revert to v7 if they need legacy support.

However, we're using pointer events in our scenario because we handle relatively complex features like detecting threshold, delay, filtering taps from drags, etc. and handling both mouse and touch events separately was a pain. In your case, if you want to maximise support, I'd advise to stick to mouse and touch events and try the snipped I sent, where you always check that a button is pressed.

Sample code would be (pseudo code):

```js
function handleMoveStart() {
  window.addEventListener('mousemove', handleMove)
  window.addEventListener('mouseup', handleMoveEnd)
  
  // start dragging logic
}

function handleMove(event) {
  if(event.buttons !== 1) {
    handleMoveEnd(event)
    return
  }

  // move dragging logic
}


function handleMoveEnd(event) {
  window.removeEventListener('mousemove', handleMove)
  window.removeEventListener('mouseup', handleMoveEnd)

  // end dragging logic
}
```

**Comment by omgovich** [comment]
Thanks guys! Gonna fix the bug today.

**Comment by omgovich** [comment]
@drcmda Just released `v4.4.4`.
As far as I see the bug is gone. Simple demo:
https://codepen.io/omgovich/pen/rNMyxoX

Thanks for the issue and help 👍

**Comment by omgovich** [comment]
Awesome floating shoe thing by the way 👍
Love how it looks and works. Really nice work!
